core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/DefaultLoadBalancingPolicy.java

```diff
@@ -132,21 +127,34 @@ public Queue<Node> newQueryPlan(@Nullable Request request, @Nullable Session ses
 
     Set<Node> allReplicas = getReplicas(request, session);
     int replicaCount = 0; // in currentNodes
+    int localRackReplicaCount = 0; // in currentNodes
+    String localRack = getLocalRack();
 
     if (!allReplicas.isEmpty()) {
 
       // Move replicas to the beginning of the plan
+      // Replicas in local rack should precede other replicas
       for (int i = 0; i < currentNodes.length; i++) {
         Node node = (Node) currentNodes[i];
         if (allReplicas.contains(node)) {
-          ArrayUtils.bubbleUp(currentNodes, i, replicaCount);
+          if (Objects.equals(node.getRack(), localRack)
+              && Objects.equals(node.getDatacenter(), getLocalDatacenter())) {
+            ArrayUtils.bubbleUp(currentNodes, i, localRackReplicaCount);
+            localRackReplicaCount++;
+          } else {
+            ArrayUtils.bubbleUp(currentNodes, i, replicaCount);
+          }
           replicaCount++;
         }
       }
 
       if (replicaCount > 1) {
-
-        shuffleHead(currentNodes, replicaCount);
+        // Shuffles only replicas that are not in local rack
```

Comment:
Why?

Reply:
@avelanarius Perhaps you can remember whether there was a specific purpose for not shuffling local rack replicas?

Reply:
As discussed with @avelanarius, it is modified so that replicas in the local rack are also shuffled separately from other replicas.